path/dynamic/dstarlite.go

```diff
@@ -0,0 +1,459 @@
+	   {45”}       else if (rhs(u) = c_old + g(v))
+	   {46”}         if (u != s_goal) rhs(u) = min s'∈Succ(u)(c(u, s') + g(s'));
+	   {47”}       UpdateVertex(u);
+	   {48”}     ComputeShortestPath()
+	*/
+	if len(changes) == 0 {
+		return
+	}
+	d.keyModifier += d.heuristic(d.last, d.s)
+	d.last = d.s
+	for _, e := range changes {
+		if e.Cost < 0 {
+			panic("D* Lite: negative edge weight")
+		}
+		cOld := d.weight(e.Edge)
+		u := d.worldNodeFor(e.From())
+		v := d.worldNodeFor(e.To())
+		d.world.SetEdge(concrete.Edge{F: u, T: v}, e.Cost)
+		if cOld > e.Cost {
+			if u.ID() != d.t.ID() {
+				u.rhs = math.Min(u.rhs, e.Cost+v.g)
+			}
+		} else if u.rhs == cOld+v.g {
+			if u.ID() != d.t.ID() {
+				u.rhs = math.Inf(1)
+				for _, t := range d.world.From(u) {
+					u.rhs = math.Min(u.rhs, d.weight(d.world.Edge(u, t))+t.(*dStarLiteNode).g)
+				}
+			}
+		}
+		d.update(u)
+	}
+	d.findShortestPath()
+}
+
+func (d *DStarLite) worldNodeFor(n graph.Node) *dStarLiteNode {
+	switch w := d.world.Node(n.ID()).(type) {
+	case *dStarLiteNode:
+		return w
+	case graph.Node:
+		panic(fmt.Sprintf("D* Lite: illegal world node type: %T", w))
+	default:
+		return &dStarLiteNode{Node: n, rhs: math.Inf(1), g: math.Inf(1)}
+	}
+}
+
+// Here returns the current location.
+func (d *DStarLite) Here() graph.Node {
+	return d.s.Node
+}
+
+// Path returns the path from the current location to the goal and the
+// weight of the path.
+func (d *DStarLite) Path() (p []graph.Node, weight float64) {
+	u := d.s
+	p = []graph.Node{u.Node}
+	for u.ID() != d.t.ID() {
+		if math.IsInf(u.rhs, 1) {
+			return nil, math.Inf(1)
+		}
+
+		min := math.Inf(1)
+		var (
+			next *dStarLiteNode
+			cost float64
+		)
+		for _, _v := range d.world.From(u) {
+			v := _v.(*dStarLiteNode)
+			w := d.weight(d.world.Edge(u, v))
+			if rhs := w + v.g; rhs < min {
+				next = v
+				min = rhs
+				cost = w
+			}
+		}
+		u = next
+		weight += cost
+		p = append(p, u.Node)
+	}
+	return p, weight
+}
+
+/*
+The pseudocode uses the following functions to manage the priority
+queue:
+
+      * U.Top() returns a vertex with the smallest priority of all
+        vertices in priority queue U.
+      * U.TopKey() returns the smallest priority of all vertices in
+        priority queue U. (If is empty, then U.TopKey() returns [∞;∞].)
+      * U.Pop() deletes the vertex with the smallest priority in
+        priority queue U and returns the vertex.
+      * U.Insert(s, k) inserts vertex s into priority queue with
+        priority k.
+      * U.Update(s, k) changes the priority of vertex s in priority
+        queue U to k. (It does nothing if the current priority of vertex
+        s already equals k.)
+      * Finally, U.Remove(s) removes vertex s from priority queue U.
+*/
+
+// key is a D* Lite priority queue key.
+type key [2]float64
+
+// less returns whether k is less than other. From ISBN:0-262-51129-0 pp476-483:
+//
+//  k ≤ k' iff k₁ < k'₁ OR (k₁ == k'₁ AND k₂ ≤ k'₂)
```

Comment:
`CTRL-K 1s` does what? It seems to have `CTRL-K` seems to have many meanings. On my machine the `AltR` `_` n sequence  (where n is a digit) gives the subscripts and `AltR` `^` n gives the superscripts.